src/util_types/mutator_set/removal_record.rs

```diff
@@ -65,6 +66,38 @@ impl AbsoluteIndexSet {
     pub fn to_array_mut(&mut self) -> &mut [u128; NUM_TRIALS as usize] {
         &mut self.0
     }
+
+    /// Split the [`AbsoluteIndexSet`] into two hash maps, one for chunks in
+    /// the inactive part of the Bloom filter and another one for chunks in the
+    /// active part of the Bloom filter.
+    ///
+    /// Returns an error if a removal index is a future value, i.e. one that's
+    /// not yet covered by the active window.
+    #[allow(clippy::type_complexity)]
+    pub fn split_by_activity(
+        &self,
+        mutator_set: &MutatorSetAccumulator,
+    ) -> Result<(HashMap<u64, Vec<u128>>, Vec<u128>), MutatorSetError> {
+        let mut inactive = indices_to_hash_map(&self.0);
```

Comment:
style: consider re-writing the for loop using iterator combinators to avoid the mutable variables.

note: @jan-ferdinand is quite good at this, if there's something tricky.

Reply:
Something along the following lines could work:

```rust
    pub fn split_by_activity(
        &self,
        mutator_set: &MutatorSetAccumulator,
    ) -> Result<(HashMap<u64, Vec<u128>>, Vec<u128>), MutatorSetError> {
        let (aw_chunk_index_min, aw_chunk_index_max) = mutator_set.active_window_chunk_interval();

        let (inactive, active): (HashMap<_, _>, HashMap<_, _>) = indices_to_hash_map(&self.0)
            .into_iter()
            .partition(|&(chunk_index, _)| chunk_index < aw_chunk_index_min);
        let active = active.into_values().flatten().collect_vec();

        if let Some((&chunk_index, _)) = inactive.iter().find(|(&i, _)| i > aw_chunk_index_max) {
            return Err(MutatorSetError::AbsoluteRemovalIndexIsFutureIndex {
                current_max_chunk_index: aw_chunk_index_max,
                saw_chunk_index: chunk_index,
            });
        }

        Ok((inactive, active))
    }
```

By the way, I agree with clippy: the return type is quite complex.

Reply:
I removed the mutability and adapted most of Ferdinand's suggestion. I'm keeping the return type though, as it in my opinion is not overly complicated (I think clippy recently made the rules for "complex" return types stricter). We don't yet have benchmarks for the mutator sets, so I can't tell which one is faster. If I had to take a guess they are equally fast though. Also: Since this all relates to values that live in RAM, this function is very unlikely to be any kind of bottleneck.

Reply:
> I think clippy recently made the rules for "complex" return types stricter

The only hits for `type_complexity` in the [clippy changelog](https://github.com/rust-lang/rust-clippy/blob/master/CHANGELOG.md) are from rust v1.59 from 2022-02-24 and rust v1.31 from 2018-12-06, both of which fix false positive, neither of which touches the default threshold for that lint. I conclude that no, clippy has not gotten more sensitive on this.

Note that a `type` alias is often sufficient. Grepping through the codepase, `HashMap<u64, Vec<u128>>` seems to be a thing. Giving that thing a name could be beneficial not just to calm clippy, but to signify what that thing is.